core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -207,22 +212,42 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   def currentStateTimestampOrDefault: Long = currentStateTimestamp.getOrElse(-1)
 
   def add(member: MemberMetadata, callback: JoinCallback = null) {
-    if (members.isEmpty)
-      this.protocolType = Some(member.protocolType)
-
-    assert(groupId == member.groupId)
-    assert(this.protocolType.orNull == member.protocolType)
-    assert(supportsProtocols(member.protocols))
+    memberPrecheck(member.memberId, member.groupId, member.protocolType, member.protocols)
 
     if (leaderId.isEmpty)
       leaderId = Some(member.memberId)
     members.put(member.memberId, member)
-    member.supportedProtocols.foreach{ case (protocol, _) => supportedProtocols(protocol) += 1 }
     member.awaitingJoinCallback = callback
     if (member.awaitingJoinCallback != null)
       numMembersAwaitingJoin += 1
   }
 
+  def addUnknownMember(memberId: String,
+                       groupId: String,
+                       protocolType: String,
+                       protocols: List[(String, Array[Byte])]) {
+    memberPrecheck(memberId, groupId, protocolType, protocols.map(_._1).toSet)
+
+    members.put(memberId, GroupMetadata.unknownMemberMetadata)
+  }
+
+  private def memberPrecheck(memberId: String,
+                             groupId: String,
+                             protocolType: String,
+                             protocols: Set[String]): Unit = {
+    if (members.isEmpty)
+      this.protocolType = Some(protocolType)
```

Comment:
From the original code, but we should probably save any mutation until we have validated the rest of the metadata.